### PR TITLE
Prewarm using lfc prefetch

### PR DESCRIPTION
**Pull request by knizhnik**
## Problem

See 
(original prewarm implementation)
https://github.com/neondatabase/neon/pull/9197
(functions for storing/restoring LFC state)
https://github.com/neondatabase/neon/pull/9587
(store prefetch results in LFC)
https://github.com/neondatabase/neon/pull/10442

## Summary of changes

This PR implements prewarm using lfc_prefetch. 
So actually it is rewritten version of  #9587 on top of #10442

### CODE REVIEW

File: pgxn/neon/pagestore_smgr.c [hunk 1]
```diff
@@ -2418,7 +2430,6 @@ neon_exists(SMgrRelation reln, ForkNumber forkNum)
 	{
 		NeonExistsRequest request = {
 			.hdr.tag = T_NeonExistsRequest,
-			.hdr.reqid = GENERATE_REQUEST_ID(),
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/pagestore_smgr.c:2505]: Because it is moved to libpagestore.c `pageserver_send` - just single place instead of specifying  it in multiple places. Prewarm also sends requests to PS and need to check responses.

File: pgxn/neon/file_cache.c [hunk 2]
```diff
@@ -524,6 +570,220 @@ lfc_init(void)
...
+ * Prewarm LFC cache to the specified state. It uses lfc_prefetch function to load prewarmed page without hoilding shared buffer lock
+ * and avoid race conditions with other backends.
+ */
+static void
+lfc_prewarm(FileCacheStateEntry* fs, size_t n_entries)
+{
+	size_t snd_idx = 0, rcv_idx = 0;
+	size_t n_sent = 0, n_received = 0;
+	int    shard_no;
+	PrewarmRequest* ring;
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:621]: It is definitely possible (I mentioned it somewhere).
It will significantly simplify code, because we do not need to work directly with PS in this case.
I have two arguments against it:
1. All this prefetching machinery from pagestore_smgr.c is not needed here (hash table, placing response page in memory,...). Unlikely this overhead can be somehow noticeable.
2. Prewarming is some special case of prefetch and can be done at larger distance. So I didn't want to limit it by `effective_io_concurrency` (ok , we can change it locally) and `readahead_buffer_size`.

But If you think that it is better to use standard prefetch mechanism for prewarming, I can change it.

File: pgxn/neon/file_cache.c [hunk 3]
```diff
@@ -524,6 +570,220 @@ lfc_init(void)
...
+	*n_entries = i;
+	return fs;
+}
+
+/*
+ * Prewarm LFC cache to the specified state. It uses lfc_prefetch function to load prewarmed page without hoilding shared buffer lock
+ * and avoid race conditions with other backends.
+ */
+static void
+lfc_prewarm(FileCacheStateEntry* fs, size_t n_entries)
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:616]: 1-2. Not sure that I completely understand you. `n_entries` is actually size of `fs` array: number of chunks which we want to prewarm. It can be smaller than size of LFC. It was per-design choice: be able to load not the whole content of LFC, which can be quite huge - terabytes, but only most frequently accessed chunks.

Also I do not understand what you mean by "present chunks". Actually prewarming means that at the beginning LFC is empty. Ok, certainly while prewarming is in progress, it can be filled by backends. And some of the pages we are going to prewarm are already present in LFC. We definitely skip them.

3. It is handled here:
```
				while (!page_server->send(shard_no, (NeonRequest *) &request)
					|| !page_server->flush(shard_no))
				{
					/* page server disconnected: all previusly sent prefetch requests are lost */
					n_sent = 0;
				}
```

[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:616]: You are right!
I still have in my head old model when there is a separate background worker process performing prewarm.
It is certainly not true now, when prewarming can be  initiated in normal backend.
I will redo it using standard prefetch mechanism.

File: pgxn/neon/file_cache.c [hunk 4]
```diff
@@ -156,11 +161,19 @@ typedef struct FileCacheControl
 
 bool lfc_store_prefetch_result;
 
+typedef struct FileCacheStateEntry
+{
+	BufferTag	key;
+	uint32		bitmap[CHUNK_BITMAP_SIZE];
+} FileCacheStateEntry;
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c]: Done

[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c]: Done

File: pgxn/neon/file_cache.c [hunk 5]
```diff
@@ -147,6 +148,10 @@ typedef struct FileCacheControl
 	uint64		time_write;		/* time spent writing (us) */
 	uint64		resizes;        /* number of LFC resizes   */
 	uint64		evicted_pages;	/* number of evicted pages */
+	uint32		prewarm_total_chunks;
+	uint32		prewarm_curr_chunk;
+	uint32		prewarmed_pages;
+	uint32		skipped_pages;
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c]: Done

File: pgxn/neon/file_cache.c [hunk 6]
```diff
@@ -512,6 +552,45 @@ lfc_init(void)
...
+							NULL,
+							&lfc_chunk_size_log,
+							MAX_BLOCKS_PER_CHUNK_LOG,
+							0,
+							MAX_BLOCKS_PER_CHUNK_LOG,
+							PGC_POSTMASTER,
+							0,
+							NULL,
+							lfc_change_chunk_size,
+							NULL);
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:590]: Yes, definitely we can.
I thought about it. But changing of this GUC is not permitted for Neon users. So it is our internal setting.
I do no see much sense in adding extra code for checking that it is power of 2 and getting logarithm 2 of it (for efficient division). But if you insist, I can easily change.

File: pgxn/neon/file_cache.c [hunk 7]
```diff
@@ -123,7 +125,7 @@ typedef struct FileCacheEntry
 	uint32		hash;
 	uint32		offset;
 	uint32		access_count;
-	uint32		state[(BLOCKS_PER_CHUNK + 31) / 32 * 2]; /* two bits per block */
+	uint32		state[CHUNK_BITMAP_SIZE * 2]; /* two bits per block */
 	dlist_node	list_node;		/* LRU/holes list node */
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:127]: Done

File: pgxn/neon/file_cache.c [hunk 8]
```diff
@@ -153,15 +166,32 @@ typedef struct FileCacheControl
...
+	PrewarmWorkerState prewarm_workers[MAX_PREWARM_WORKERS];
 } FileCacheControl;
 
-bool lfc_store_prefetch_result;
+#define FILE_CACHE_STATE_MAGIC 0xfcfcfcfc
+
+typedef struct FileCacheState
+{
+	uint32		magic;
+	uint32		n_chunks;
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:175]: To estimate progress in pages we need to calculate total number of marked pages in all blocks and complicates logic of skipping chunk (need to calculate pages in this chunk).
Certainly can be done.
But I am not sure that it will somehow significantly increase progress reporting precision and that somebody will ever monitor this progress...
So I do not think that extra efforts cost the money.

File: pgxn/neon/file_cache.c [hunk 9]
```diff
@@ -525,6 +604,200 @@ lfc_init(void)
...
+				snd_idx += 1 << fcs_chunk_size_log;
+			}
+			else
+			{
+				if (BITMAP_ISSET(bitmap, snd_idx))
+				{
+					tag = fcs->chunks[snd_idx >> fcs_chunk_size_log];
+					tag.blockNum += snd_idx & ((1 << fcs_chunk_size_log) - 1);
+					(void)prefetch_register_bufferv(tag, NULL, 1, NULL, true);
+					n_sent += 1;
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c]: Good catch.
Actually we don't.
I always assumed that prewarm is done on cold system ands LFC is almost empty.
But I agree with you that it is better to add check here to avoid redundant requests to PS.

File: pgxn/neon/file_cache.c [hunk 10]
```diff
@@ -525,6 +604,200 @@ lfc_init(void)
...
+				/* Skip chunks processed by other workers */
+				rcv_idx += 1 << fcs_chunk_size_log;
+				continue;
+			}
+
+			/* Locate next block to prefetch */
+			while (!BITMAP_ISSET(bitmap, rcv_idx))
+			{
+				rcv_idx += 1;
+			}
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:771]: I think we can do if it will be visible in any profile.
But I have very serious doubts that this overhead can be somehow measurable.
Iterating through even megabyte bitmap will take milliseconds. It is not comparable with time of load of even single page.

Optimizations makes sense if bitmap is mostly sparse. But it can be very sparse because empty chunks are not included in LFC state.

File: pgxn/neon/file_cache.c [hunk 11]
```diff
@@ -1684,3 +1982,83 @@ approximate_working_set_size(PG_FUNCTION_ARGS)
...
+	if (fcs != NULL)
+	{
+		size_t size_in_bytes = FILE_CACHE_STATE_SIZE(fcs->n_chunks);
+		bytea* res = (bytea*)palloc(VARHDRSZ + size_in_bytes);
+
+		SET_VARSIZE(res, VARHDRSZ + size_in_bytes);
+		memcpy(VARDATA(res), fcs, size_in_bytes);
+		pfree(fcs);
+
+		PG_RETURN_BYTEA_P(res);
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:2002]: I think about it but decided to make interface simpler in price of extra copy.
But after your comment I noticed that creating varlena types is quite popular in Postgres,
so I made this change.

File: pgxn/neon/file_cache.c [hunk 12]
```diff
@@ -133,13 +135,24 @@ typedef struct FileCacheEntry
...
 #define CV_WAIT_TIMEOUT	10
 
+#define MAX_PREWARM_WORKERS 8
+
+typedef struct PrewarmWorkerState
+{
+	uint32		total_chunks;
+	uint32		curr_chunk;
+	uint32		prewarmed_pages;
+	uint32		skipped_pages;
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:144]: Sorry, may be I do not understand you.

We currently  have function `prewarm_local_cache(state bytea, worker_id integer default 0, n_workers integer default 1)`. So it contains `worker_id`. If we want to perform upload from different backends, we should call this function with different `worker_id`.

You can check state of each worker using `get_prewarm_info(worker_id integer default 0,...)` function by specifying proper `worker_id`.

We do not need PIDs in this schema, do we?

File: pgxn/neon/file_cache.c [hunk 13]
```diff
@@ -133,13 +135,24 @@ typedef struct FileCacheEntry
...
+} PrewarmWorkerState;
+
 typedef struct FileCacheControl
 {
 	uint64		generation;		/* generation is needed to handle correct hash
 								 * reenabling */
 	uint32		size;			/* size of cache file in chunks */
 	uint32		used;			/* number of used chunks */
 	uint32		used_pages;		/* number of used pages */
+	uint32		pinned;			/* number of pinned chunks */
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:154]: Because it is accessed under LFC lock.

File: pgxn/neon/file_cache.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:177]: Done

File: pgxn/neon/neon--1.5--1.6.sql [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+LANGUAGE C STRICT
+PARALLEL SAFE;
+
+CREATE FUNCTION get_local_cache_state(max_chunks integer default null)
+RETURNS bytea
+AS 'MODULE_PATHNAME', 'get_local_cache_state'
+LANGUAGE C
+PARALLEL UNSAFE;
+
+CREATE FUNCTION prewarm_local_cache(state bytea, worker_id integer default 0, n_workers integer default 1)
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/neon--1.5--1.6.sql:15]: How?
I do not see any any reasonable way to generate such ID and determine indented number of background workers.
So I prefer to redirect this decision to control plane and keep prewarm API as simple and straightforward as possible. This is why I also do not want to somehow invoke worker PID in it.

File: pgxn/neon/file_cache.c [hunk 15]
```diff
@@ -525,6 +604,200 @@ lfc_init(void)
...
+		return;
+	}
+
+	if (fcs->magic != FILE_CACHE_STATE_MAGIC)
+	{
+		elog(ERROR, "LFC: Invalid file cache state magic: %X", fcs->magic);
+	}
+
+	fcs_chunk_size_log = fcs->chunk_size_log;
+	if (fcs_chunk_size_log > MAX_BLOCKS_PER_CHUNK)
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:710]: Why?
I want to have strict upper boundary for chunk size to be able to allocate it statically.
This is why all this ` MAX_BLOCKS_PER_CHUNK` constant was introduced.
If it is not changed, then this check actually do not have sense.
But we can not somehow enforce it.
So I think that it is better to check it, isn't it?

File: pgxn/neon/file_cache.c [hunk 16]
```diff
@@ -525,6 +604,200 @@ lfc_init(void)
...
+	if (lfc_ctl->limit <= lfc_ctl->size)
+	{
+		elog(LOG, "LFC: skip prewarm because LFC is already filled");
+		LWLockRelease(lfc_lock);
+		return;
+	}
+	vacant_blocks = (size_t)(lfc_ctl->limit - lfc_ctl->size) << lfc_chunk_size_log;
+	if (n_entries > (vacant_blocks >> fcs_chunk_size_log))
+	{
+		n_entries = vacant_blocks >> fcs_chunk_size_log;
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:729]: Yes, if LFC is already partly filled with data, then this check doesn't make a lot of sense.
But my assumption was that it makes sense to prewarm only cold compute - with empty LFC.
And as far as LFC size as replica can be smaller than on source node, I prefer to prevent to useless job of storing in LFC pages which will be evicted by more recently accessed pages during prewarm.

So I think that such kind of check is absolutely needed.
I will be pleased if you can propose more precise check.

File: pgxn/neon/file_cache.c [hunk 17]
```diff
@@ -525,6 +604,200 @@ lfc_init(void)
...
+	{
+		elog(ERROR, "LFC: Invalid file cache state magic: %X", fcs->magic);
+	}
+
+	fcs_chunk_size_log = fcs->chunk_size_log;
+	if (fcs_chunk_size_log > MAX_BLOCKS_PER_CHUNK)
+	{
+		elog(ERROR, "LFC: Invalid chunk size log: %u", fcs->chunk_size_log);
+	}
+	n_entries = Min(fcs->n_chunks, lfc_prewarm_limit);
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:714]: Thank you for noticing it.
Fixed.

**Comment by knizhnik** [comment]
> @knizhnik there are still 4 (apparently collapsed-by-default) conversations you seem to have missed, or at least haven't replied to:
> 
> * [Prewarm using lfc prefetch #10591 (comment)](https://github.com/neondatabase/neon/pull/10591#discussion_r2033955217)
> * [Prewarm using lfc prefetch #10591 (comment)](https://github.com/neondatabase/neon/pull/10591#discussion_r2035575184)
> * [Prewarm using lfc prefetch #10591 (comment)](https://github.com/neondatabase/neon/pull/10591#discussion_r2035907622)
> * [Prewarm using lfc prefetch #10591 (comment)](https://github.com/neondatabase/neon/pull/10591#discussion_r2035908603)

Sorry, for some reasons I have not noticed them.
Replied.

File: pgxn/neon/file_cache.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:710]: Ugh:(
Sorry for may inattention.
Fixed

File: pgxn/neon/file_cache.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:594]: I have changed GUC to "neon.chunk_size" and it specifies number of blocks per chunk.
I agree with you that it will be more convenient even if it is only for internal use.

File: test_runner/regress/test_lfc_prewarm.py [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+    cur.execute("select sum(pk) from t")
+    assert cur.fetchall()[0][0] == n_records * (n_records + 1) / 2
+
+    assert prewarm_info[1] > 0
+
+
+@pytest.mark.skipif(not USE_LFC, reason="LFC is disabled, skipping")
+def test_lfc_prewarm_under_workload(neon_simple_env: NeonEnv):
+    env = neon_simple_env
+    n_records = 1000000
```
Comment (myrrc) [inline test_runner/regress/test_lfc_prewarm.py:63]: Can we lower down number of records in tests? They take around 60 seconds each

File: test_runner/regress/test_lfc_prewarm.py [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+            cur.execute("select pg_reload_conf()")
+            cur.execute("alter system set neon.file_cache_size_limit='1GB'")
+            cur.execute("select pg_reload_conf()")
+            cur.execute("select prewarm_local_cache(%s)", (lfc_state,))
+            n_prewarms += 1
+        log.info(f"Number of prewarms: {n_prewarms}")
+
+    workload_threads = []
+    for _ in range(n_threads):
+        t = threading.Thread(target=workload)
```
Comment (myrrc) [inline test_runner/regress/test_lfc_prewarm.py:114]: Can we use a https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor for submitting tasks and waiting for them to return?

File: test_runner/regress/test_lfc_prewarm.py (hunk 18 above)
Comment (knizhnik) [reply test_runner/regress/test_lfc_prewarm.py:63]: I think it is not because number of records, but because of this sleep:
```
 time.sleep(100)
```

Ok, I will reduce it.

File: test_runner/regress/test_lfc_prewarm.py (hunk 19 above)
Comment (knizhnik) [reply test_runner/regress/test_lfc_prewarm.py:114]: I do not see much sense in it. I need to spawn N workload threads and 1 prewarm thread.
Moreover I need to run workload threads during some time (100 seconds).
I do not know how to implement it using ThreadPoolExecutor and do not see some huge advantages except removing ~10 lines of code (loop of launching threads and waiting their termination).

Comment (myrrc) [reply test_runner/regress/test_lfc_prewarm.py:114]: Up to you

File: pgxn/neon/file_cache.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c]: Ok, I think that there can really be cases when prewarm is interrupted and then restarted.
In this case it will be useful to have more flexible criteria of prewarm completion.
So I have implemented your suggestion.

File: pgxn/neon/file_cache.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:143]: I do not understand how this comment  https://github.com/neondatabase/neon/pull/10591#discussion_r2042074161  is related to the problem of spawning two concurrent prewarms with the same worker_id.
But I have addressed complaints in this comment.

File: pgxn/neon/neon--1.5--1.6.sql (hunk 14 above)
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/neon--1.5--1.6.sql:15]: Hmmm... I do not see any relation between worker_id and shard_id. And IMHO I do not think that there should be some relation between them. Number of workers is mostly limited by network capacity of compute and its CPU utilisation. We should start as many workers as needed to provide fastest prewarm. I do not think that each worker should be pinned to some particular shard. I do not want to add this dependency and let LFC know something about shard hash function.

We round robin chunks between workers. Chunks may belong to different shards. And prewarm can be performed in parallel.  The only restriction now is that you should specify number of workers from the very beginning. It is not possible to increase number of workers until we find out that adding more workers doesn't increase total prewarming speed. 

May be I do not understand your complaint. If so, can you please provide more details: what is wring with current schema and what do you propose?

[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/neon--1.5--1.6.sql:15]: I wonder how it is possible to prepare 2 sets of LFC prewarm data and especially why may we need it?
State of LFC corresponds to some snapshot of compute node.
Certainly we can take multiple snapshots. But only last one of them is most relevant. 

So the model I have in my mind and which I tried to represent in this prewarm API (if two functions can be called "API") is the following:

1. There is just one most actual stored LFC state
2. To improve prewarm speed we can scatter prewarming work between several workers. In this case chunks from stored prefetch state are scattered between workers. To support such parallel prewarm it is necessary to specify `worker_id` and `n_workers`.
3.

File: test_runner/regress/test_lfc_prewarm.py [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+    assert prewarm_info[0] > 0 and prewarm_info[0] == prewarm_info[1]
+
+    cur.execute("select sum(pk) from t")
+    assert cur.fetchall()[0][0] == n_records * (n_records + 1) / 2
+
+    assert prewarm_info[1] > 0
+
+
+@pytest.mark.skipif(not USE_LFC, reason="LFC is disabled, skipping")
+def test_lfc_prewarm_under_workload(neon_simple_env: NeonEnv):
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply test_runner/regress/test_lfc_prewarm.py:61]: Sorry, I do not completely understand what you mean.
Prewarm process is using prefetch and prefetch is updating state of HLL.
It is IMHO intended behaviour: we are restoring seed state of LFC and it seems to be desired that working set estimation should be the same as at the point when this state was captured. 

What di you suggest?
Not to update HLL during prewarm?

File: pgxn/neon/file_cache.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:153]: When we capture LFC state we are using LRU list - to prewarm first most frequently accessed entries. Certainly LRU may be not the best policy for it. But it is IMHO better than nothing.

The problem is that pinned entries are not present in LRU list. So they are not included in captured LFC state.
I do not think that it is a bug problem: entry is pinned for quite short time and o number of such "missed" entries should be small. 

So actual  number of grabbed entries is `used-pinned`.
Also information about pinned p[ages seems to be useful regardless of prewarm. This is why it is reported by `lfc_get_state`.

File: test_runner/regress/test_lfc_prewarm.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (knizhnik) [reply test_runner/regress/test_lfc_prewarm.py:61]: > Prewarming is not an actual user workload, 

Yes, it is not user workload. But as I wrote above it "restores" user workload. So of WS estimation before restart was i.e. 100Gb, then after prewarm WS estimation should be also 100Gb.
So IMHPO it is intended behaviour.

[the remark this answers is not in the document]
Comment (knizhnik) [reply test_runner/regress/test_lfc_prewarm.py:61]: Well... Perform is not necessarily loading all pages. You can limit number of entries when capture LFC state and when perform prewarm.
But it  generally it is assumed that LFC contains working set (or part of working set if it doesn't fit in LFC).
Certainly it may be not true: some entries may be loaded in LFC long time ago and not accessed since this time. 

In any case - I think that there is no better alternative. You suggestion "not to update HLL during perform" and " make the user's workload determine the working set estimation" will lead to the situation when estimated WS will be very small and autoscaler tries to scale node down, shaking LFC cache and throwing away everything we have just prewarmed.

I think that we should make an assumption that captures LFC state actually represents user's workload. And we should update HLL as if such pages are requested by users.

File: pgxn/neon/file_cache.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:153]: Sorry, but I do not understand what such test should check?
What can be considered as "significant leak". Assume that we have very small LFC - just 100 chunks.
And we open 100 sessions each of which randomly access one of the pages. Definitely in this case it is possible that 100% of LFC entries are pinned. But in practice it should never happen: LFC size his expected to be much larger than maximal number of backends. In this case even in the worst situation when each backend accesses its own page, pins it and starts waiting response from PS, number of pinned pages will be relatively small. There is IMHO not so much sense in writing test for it: maximal number of pinned pages can not exceed number of backends in any case.

File: test_runner/regress/test_lfc_prewarm.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (knizhnik) [reply test_runner/regress/test_lfc_prewarm.py:61]: > True, but LFC can also contain non-WS pages when the WS is smaller than the LFC. 

Certainly, I also mentioned this case. In this case our estimation of WS will be too pessimistic preventing autoscaler from reducing LFC size. But do we really want autoscaler to throw away pages which we just loaded as part of prewarm? It will be extremely strange and inefficient approach.

The better way may be to be more smart when capturing LFC state and for example do not include rarely accessed pages in it. It may be not so trivial (requires storing timestamp or access counter) but possible. 
But frankly speaking I think that overkill. WS either fit in LFC, in this case we should just load it, doesn't matter how frequently this pages are accessed, ether do not fit in it - in this case we need to load least recently accessed pages.

[the remark this answers is not in the document]
Comment (knizhnik) [reply test_runner/regress/test_lfc_prewarm.py:61]: > Do you have an example for that? AFAIK we don't have issues with that right now, and it definitely won't be any worse once prewarming is in place.

As far as I remember  we had such issue and it was discussed at autoscaler meeting. I not sure if some solution was proposed (rather than prohibit LFC downscale until some time is passed since node startup).

File: pgxn/neon/file_cache.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:153]: But we have test_lfc_resize.which doesn't exactly it.
I can add printing of pinned entries counter at the end of the test to make sure that it is zero.
Will it be enough?

[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:153]: I have added this check to test_lfc_prewarm.py tests

File: pgxn/neon/neon--1.5--1.6.sql (hunk 14 above)
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/neon--1.5--1.6.sql:15]: Yes, user can do it.
My opinion is that we should not prohibit user to manually perform prewarm but should not try to optimise it for any use cases which user can imagine.

**Comment by knizhnik** [comment]
I have implemented prefetch using background workers.
@MMeent  please check if it satisfies your requirements.

File: pgxn/neon/file_cache.c [hunk 21]
```diff
@@ -521,6 +627,287 @@ lfc_init(void)
...
+
+	/* Calculate total number of pages to be prewarmed */
+	lfc_ctl->total_prewarm_pages = fcs->n_pages;
+
+	seg = dsm_create(fcs_size, 0);
+	memcpy(dsm_segment_address(seg), fcs, fcs_size);
+	lfc_ctl->prewarm_lfc_state_handle = dsm_segment_handle(seg);
+
+	/* Spawn background workers */
+	for (uint32 i = 0; i < n_workers; i++)
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:766]: I think about it. Rejected this idea because:
1. Error handling. Now if some error happen in prewarm, if affect only one BGW, but not coordinator. It is still a question how to better inform coordinator about such errors. `WaitForBackgroundWorkerShutdown` returns `BgwHandleStatus` which doesn't include error state.
2. Simplify logic: there is coordinator and there are workers.Trying to combine this two roles in one process is not "elegant".
3. If we will actually use many parallel workers (and I expect almost linear reduction of prewarm time), then optimising this corner case (one worker) seems to be very useful and important.

File: pgxn/neon/file_cache.c [hunk 22]
```diff
@@ -521,6 +627,287 @@ lfc_init(void)
...
+		lfc_ctl->prewarm_workers[i].handle = NULL;
+	}
+	dsm_detach(seg);
+
+	LWLockAcquire(lfc_lock, LW_EXCLUSIVE);
+	lfc_ctl->prewarm_active = false;
+	LWLockRelease(lfc_lock);
+}
+
+PGDLLEXPORT void lfc_prewarm_main(Datum main_arg);
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:805]: Done

File: pgxn/neon/file_cache.c [hunk 23]
```diff
@@ -119,23 +125,34 @@ typedef struct FileCacheEntry
...
 #define N_COND_VARS 	64
 #define CV_WAIT_TIMEOUT	10
 
+#define MAX_PREWARM_WORKERS 8
+
+typedef struct PrewarmWorkerState
+{
+	uint32		prewarmed_pages;
+	uint32		skipped_pages;
+	BackgroundWorkerHandle *handle;
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:145]: After some thinking - remove handle from worker state and replace it with completion time.

File: pgxn/neon/file_cache.c [hunk 24]
```diff
@@ -521,6 +627,287 @@ lfc_init(void)
...
+	memset(lfc_ctl->prewarm_workers, 0, n_workers*sizeof(PrewarmWorkerState));
+
+	LWLockRelease(lfc_lock);
+
+	/* Calculate total number of pages to be prewarmed */
+	lfc_ctl->total_prewarm_pages = fcs->n_pages;
+
+	seg = dsm_create(fcs_size, 0);
+	memcpy(dsm_segment_address(seg), fcs, fcs_size);
+	lfc_ctl->prewarm_lfc_state_handle = dsm_segment_handle(seg);
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:765]: Main argument is worker_id.
What is the problem with passing it through shame? Autoprewarm does the same.

File: pgxn/neon/file_cache.c [hunk 25]
```diff
@@ -1438,7 +1846,12 @@ neon_get_lfc_stats(PG_FUNCTION_ARGS)
 			break;
 		case 8:
 			key = "file_cache_chunk_size_pages";
-			value = BLOCKS_PER_CHUNK;
+			value = lfc_blocks_per_chunk;
+			break;
+		case 9:
+			key = "file_cache_pinned";
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:1863]: Done

File: pgxn/neon/file_cache.c [hunk 26]
```diff
@@ -521,6 +627,287 @@ lfc_init(void)
...
+	for (uint32 i = 0; i < n_workers; i++)
+	{
+		BackgroundWorker worker = {0};
+
+		worker.bgw_flags = BGWORKER_SHMEM_ACCESS;
+		worker.bgw_start_time = BgWorkerStart_ConsistentState;
+		worker.bgw_restart_time = BGW_NEVER_RESTART;
+		strcpy(worker.bgw_library_name, "neon");
+		strcpy(worker.bgw_function_name, "lfc_prewarm_main");
+		sprintf(worker.bgw_name, "LFC prewarm worker %d", i+1);
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/file_cache.c:777]: Done

**Comment by knizhnik** [comment]
Splitted in 3 PRs:
https://github.com/neondatabase/neon/pull/11740
https://github.com/neondatabase/neon/pull/11741
https://github.com/neondatabase/neon/pull/11742
